### PR TITLE
Log exceptions that occurred while indexing using Util.log(...)

### PR DESCRIPTION
**Pull request by fedejeanne**
## What it does
Log exceptions in package `org.eclipse.jdt.internal.core.search.indexing` using `Util.log(...)` instead of `printStackTrace`. In most cases, the logging will only take place if `JobManager.VERBOSE == true` which makes these changes _opt-in_.

Contributes to https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1303

### Activate the `VERBOSE`-mode int `JobManager` 
* Add this to the *VM-Args* of the _Run/Debug Configuration_ of **sdk.product**: `-Dosgi.debug`
* Create/Edit `<ECLIPSE_ROOT>\.options` and add: 
```properties
# Turn on debugging
# This needs to be "true" in order to enable other DEBUG/VERBOSE flags below
org.eclipse.jdt.core/debug=true

#JobManager.VERBOSE
org.eclipse.jdt.core/debug/indexmanager=true
```

## A small hack to test it: randomly throw a `RuntimeException`
* Add this at the bottom of `org.eclipse.jdt.internal.core.search.indexing.SourceIndexer.indexDocument()` :
```java
@Override
public void indexDocument() {
	...
	
	try {
		// Add this line
		if (Math.random() > 0.5) throw new RuntimeException("This should be logged into the log file");
		
		...
	} catch (Exception e) {
		if (JobManager.VERBOSE) {
			Util.log(e);
		}
	}
}
```
* Run **sdk.product**.
* Open/Create a java project.
* Create/Edit some class, save it, repeat.

You should start seeing errors like this one in the log file:
```java
!ENTRY org.eclipse.jdt.core 4 0 2023-08-22 10:10:03.886
!MESSAGE Unexpected internal error
!STACK 0
java.lang.RuntimeException: This should be logged into the log file
	at org.eclipse.jdt.internal.core.search.indexing.SourceIndexer.indexDocument(SourceIndexer.java:116)
	at org.eclipse.jdt.internal.core.search.JavaSearchParticipant.indexDocument(JavaSearchParticipant.java:83)
	at org.eclipse.jdt.internal.core.search.indexing.IndexManager.indexDocument(IndexManager.java:674)
	at org.eclipse.jdt.internal.core.search.indexing.IndexManager$2.execute(IndexManager.java:1287)
	at org.eclipse.jdt.internal.core.search.processing.JobManager.indexerLoop(JobManager.java:527)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

... and also entries like this one in the **Error Log** view:
image

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public int hashCode() {
 	@Override
 	public boolean execute(IProgressMonitor progressMonitor) {
 
-		if (this.isCancelled || progressMonitor != null && progressMonitor.isCanceled()) return true;
+		if (this.isCancelled || ((progressMonitor != null) && progressMonitor.isCanceled())) return true;
```
Comment (HeikoKlare) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java:95]: Is this an auto-formatting result? Syntactically, the brackets are unnecessary and at least the ones put around the progressMonitor check reduce readability.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java [hunk 2]
```diff
@@ -140,8 +140,7 @@ public boolean execute(IProgressMonitor progressMonitor) {
 						file = org.eclipse.jdt.internal.core.util.Util.toLocalFile(location, progressMonitor);
 					} catch (CoreException e) {
 						if (JobManager.VERBOSE) {
-							org.eclipse.jdt.internal.core.util.Util.verbose("-> failed to index " + location.getPath() + " because of the following exception:"); //$NON-NLS-1$ //$NON-NLS-2$
-							e.printStackTrace();
+							org.eclipse.jdt.internal.core.util.Util.log(e,"-> failed to index " + location.getPath() + " because of the following exception:"); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (HeikoKlare) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java:143]: Maybe use static imports for `log` and the existing `verbose`?

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java [hunk 3]
```diff
@@ -318,6 +317,6 @@ public String toString() {
 	}
 
 	protected boolean hasPreBuiltIndex() {
-		return !this.forceIndexUpdate && (this.indexFileURL != null && this.indexFileURL.exists());
+		return !this.forceIndexUpdate && ((this.indexFileURL != null) && this.indexFileURL.exists());
```
Comment (HeikoKlare) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java:320]: See above: is this an auto-formatting result? Maybe remove at least the first added pair of brackets?

**Review by HeikoKlare — COMMENTED** [review]
I would appreciate further explanation on why the exceptions have to be logged instead of only printing them in verbose mode. It looks to me as if the code was designed to gracefully handle these exceptions, i.e., that the exception are actually not _exceptional_ but can occur in "valid" scenarios. Then it would make sense that they did not appear in logs and are only printed if someone executes the code in debug mode with verbose output.
I do also not understand why it was insufficient to print the errors in `verbose` mode rather than logging them.  I do not see any places in the code where acticating the `verbose` mode logs things, but all places I found simply use the `verbose()` method to print to console.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java (hunk 1 above)
Comment (fedejeanne) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java]: I was an auto-formatting setting, thank you. I reverted these changes

**Comment by iloveeclipse** [comment]
There is a difference between not properly logged exceptions (like try/catch/printStackTrace()) and **additional debug** information logged. It makes no sense to add **tracing** information to the log, it should be added to **debug trace**.
This PR mixes everything together.

See DebugOptionsListener usage. JDT currently doesn't use that fully (it uses it to only change trace flags, but it doesn't use DebugTrace.
To do so, org.eclipse.jdt.internal.core.JavaModelManager.registerDebugOptionsListener(BundleContext) should create debug trace in optionsChanged(DebugOptions) callback via options.newDebugTrace() and use that DebugTrace instead of system out in all places like org.eclipse.jdt.internal.core.util.Util.verbose(String).

So if you are interested in this, I would propose to keep the current PR to **only** address places like in the AbstractIndexer where exception handling is done in inappropriate way, and create another PR that would redirect tracing output to DebugTrace.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AbstractIndexer.java [hunk 4]
```diff
@@ -209,7 +210,7 @@ public void addMethodDeclaration(
 							typeModifiers,
 							extraFlags));
 		} catch (Exception e) {
-			e.printStackTrace();
+			Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AbstractIndexer.java:213]: This is OK.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java [hunk 5]
```diff
@@ -103,8 +103,7 @@ public boolean visit(IResourceProxy proxy) /* throws CoreException */{
 			}
 		} catch (CoreException e) {
 			if (JobManager.VERBOSE) {
-				Util.verbose("-> failed to add " + this.folderPath + " to index because of the following exception:", System.err); //$NON-NLS-1$ //$NON-NLS-2$
-				e.printStackTrace();
+				Util.log(e, "-> failed to add " + this.folderPath + " to index because of the following exception:"); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java:106]: This is not OK, see my explanation above.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java (hunk 5 above)
Comment (fedejeanne) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java:106]: If I understand you correctly, 
* All calls to `Util.verbose("...")` and `Util.verbose("...", System.err)` are **debug information** and they should be done via the `DebugTrace` (in another PR)
* All calls to `e.printStackTrace()` are meant to be logged and they should be replaced by a call to `Util.log(e)` in this PR.

?

Comment (iloveeclipse) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java:106]: Almost correct.

> All calls to e.printStackTrace()

All calls to e.printStackTrace() (or System.(out|err).print* etc) in context of "usual" error handling where no "Verbose/Debug/Trace" etc code is involved and where we (still) have platform running (there are cases on shutdown where no bundle logging is possible).

Comment (fedejeanne) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java]: Thank you, I hope I got it right this time :)

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java [hunk 6]
```diff
@@ -1800,7 +1800,7 @@ public boolean execute(IProgressMonitor progress) {
 					if (JobManager.VERBOSE) {
 						Util.verbose("-> failed to update meta index for index " + indexFile.getName() //$NON-NLS-1$
 								+ " because of the following exception:"); //$NON-NLS-1$
-						e.printStackTrace();
+						Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/IndexManager.java:1803]: This is strange one - no idea why it expects a cause from IOException, is a single IOException not good enough?

Same as above, in this PR either don't touch code functionality inside "debug" or "verbose" blocks, or refactor to log some of the exceptions.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java [hunk 7]
```diff
@@ -114,7 +115,7 @@ public void indexDocument() {
 				this.document.requireIndexingResolvedDocument();
 		} catch (Exception e) {
 			if (JobManager.VERBOSE) {
-				e.printStackTrace();
+				Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java:118]: Same as above, in this PR either don't touch code functionality inside "debug" or "verbose" blocks, or refactor to log some of the exceptions.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java [hunk 8]
```diff
@@ -172,7 +173,7 @@ public void resolveDocument() {
 			this.cud.resolve();
 		} catch (Exception e) {
 			if (JobManager.VERBOSE) {
-				e.printStackTrace();
+				Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java:176]: Same as above, in this PR either don't touch code functionality inside "debug" or "verbose" blocks, or refactor to log some of the exceptions.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java [hunk 9]
```diff
@@ -260,7 +261,7 @@ public void indexResolvedDocument() {
 			}
 		} catch (Exception e) {
 			if (JobManager.VERBOSE) {
-				e.printStackTrace();
+				Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexer.java:264]: Same as above, in this PR either don't touch code functionality inside "debug" or "verbose" blocks, or refactor to log some of the exceptions.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexerRequestor.java [hunk 10]
```diff
@@ -525,7 +528,7 @@ public void popTypeName() {
 		try {
 			this.enclosingTypeNames[-1] = null;
 		} catch (ArrayIndexOutOfBoundsException e) {
-			e.printStackTrace();
+			Util.log(e);
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexerRequestor.java:531]: This code can't be normally executed, the code inside try will **always** throw an exception, the stack trace was only for debug purpose. So please simply revert this change or replace entire "try/catch" with `new Exception().printStackTrace()`.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java (hunk 5 above)
Comment (iloveeclipse) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java]: > Thank you, I hope I got it right this time :)

Not really, see below.

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java [hunk 11]
```diff
@@ -104,7 +104,7 @@ public boolean visit(IResourceProxy proxy) /* throws CoreException */{
 		} catch (CoreException e) {
 			if (JobManager.VERBOSE) {
 				Util.verbose("-> failed to add " + this.folderPath + " to index because of the following exception:", System.err); //$NON-NLS-1$ //$NON-NLS-2$
-				e.printStackTrace();
+				Util.log(e);
```
[the remark this answers is not in the document]
Comment (fedejeanne) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddFolderToIndex.java:107]: Good point. I think every call to `e.printStackTrace()` that's inside an `if (JobManager.VERBOSE)` was never intended to be logged in the error logs but in the **tracing** logs. I will leave them out of this PR and address them in this other PR instead: https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1374.

To be clear: this means that I will **not** add more errors to the error logs. I will just log the errors in the logs they were originally intended for _i.e._:
* VERBOSE-errors --> **trace.log** (https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1374)
* Non-VERBOSE-errors --> **error.log** (this PR)

File: org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexerRequestor.java (hunk 10 above)
Comment (fedejeanne) [reply org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/SourceIndexerRequestor.java:531]: I took care of this one in https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1374/files#diff-d9c0f9e135f1b04312aa1ed45b5af71e2f7d2822751572c81244b232216241c3

**Comment by fedejeanne** [comment]
> LGTM, thanks.

Thank **you** for being so thorough! 💪

**Comment by iloveeclipse** [comment]
I see `504 Gateway Time-out` issues with jenkins. The failed test runs locally.

**Comment by iloveeclipse** [comment]
Created https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/3672 for jenkins issue...

**Comment by iloveeclipse** [comment]
jenkins build was OK, see https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-1304/5/console
```
15:45:02  [INFO] ------------------------------------------------------------------------
15:45:02  [INFO] BUILD SUCCESS
15:45:02  [INFO] ------------------------------------------------------------------------
15:45:02  [INFO] Total time:  22:40 min
15:45:02  [INFO] Finished at: 2023-09-13T13:44:22Z
15:45:02  [INFO] ------------------------------------------------------------------------
15:45:02  [WARNING] 
15:45:02  [WARNING] Plugin validation issues were detected in 3 plugin(s)
15:45:02  [WARNING] 
15:45:02  [WARNING]  * org.apache.maven.plugins:maven-site-plugin:3.12.1
15:45:02  [WARNING]  * org.codehaus.mojo:build-helper-maven-plugin:3.3.0
15:45:02  [WARNING]  * org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.2
15:45:02  [WARNING] 
15:45:02  [WARNING] For more or less details, use 'maven.plugin.validation' property with one of the values (case insensitive): [BRIEF, DEFAULT, VERBOSE]
15:45:02  [WARNING] 
Resuming build at Wed Sep 13 10:47:04 EDT 2023 after Jenkins restart
Waiting for reconnection of centos-8-dbzkh before proceeding with build
Timeout set to expire in 3 min 2 sec
Still paused
Cancelling nested steps due to timeout
Body did not finish within grace period; terminating with extreme prejudice
504 Gateway Time-out
```
